### ISSUE TITLE
False positive `private-assoc-item` error

### ISSUE THREAD

**Issue by EFanZh** [rust-lang/rust-analyzer#14046]
- **rust-analyzer version**: 0.3.1377-standalone (daa0138e4 2023-01-21)
- **rustc version**: 1.67.0 (fc594f156 2023-01-24)

Create a project with the following code in `src/main.rs`:

```rust
mod foo {
    pub struct Foo;

    const _: () = {
        impl Foo {
            pub(crate) fn bar(self) {}
        }
    };
}

fn main() {
    foo::Foo.bar()
}
```

`cargo run` executes successfully, but I get the following error in the IDE:

<img width="295" alt="image" src="https://user-images.githubusercontent.com/2160692/215303205-2f39485f-47c1-49dc-8036-2fa142736cd9.png">
